Create Python Extension Source/Connector SDK (#312)
Vantiq/vantiq-extension-sources: extpsdk/README.md

## Repository Contents

This repository contains the SDK for building Vantiq connectors (also known as extension sources)
using Python. At a high level, a connector is configured and then run, responding or sending messages to messages the Vantiq server. Each connector acts as a _source_ in a Vantiq namespace.

You will need credentials to a Vantiq server to use this SDK.  Please contact your Vantiq system administrator to obtain these credentials.

## Overview of a Connector

For details about the overall structure of a connector and the protocol used between the connector and the Vantiq source, please see [here](../README.md).  

> Each running instance of this code represents a set of connectors that connect to a single Vantiq namespace. The set of of sources for the connector & the connection information is found in the server config file.  We will describe the details later.

Reviewer: ```suggestion
Each running instance of this code represents a set of connectors that connect to a single Vantiq namespace. The set of sources for the connector & the connection information is found in the server config file.  We will describe the details later.
```

    * `query` -- when the Vantiq server queries the source, that query is delivered to the `query` handler.  The query handler can return information or errors to the application running on the server.
    * `close` -- when the connection to the Vantiq server ends, the `close` handler is called, giving the connector a chance to clean up resources.

The connector runtime will maintain the connection, reconnecting as needed.

The following sections describe the use of this SDK to construct a connector.

## Building a Connector

Running this code required Python 3.8 or later.

To use the SDK, install it into your Python environment.  Using `pip`, 

```commandline
pip install vantiqconnectorsdk
```

Other Python environment tools may use different commands.

Once it is installed, you can import it.

```python

from vantiqconnectorsdk import VantiqConnector, VantiqConnectorSet, VantiqConnectorException, VantiqSourceConnection
```

The SDK is built using _asyncio_.  For details about working with _asyncio_, please see [Python `asyncio` documentation](https://docs.python.org/3/library/asyncio.html).

The Vantiq Connector SDK includes the following items:

> * VantiqConnector -- a set of conststants defined as part of the interface.

Reviewer: ```suggestion
* VantiqConnector -- a set of constants defined as part of the interface.
```

Might also be clearer if you move this a few lines down to where the constants are listed.

	
## <a name="VantiqSourceConnection" id="sourceConnection"></a>Using ExtensionWebSocketClient
Every `VantiqSourceConnection` deals with a single source via its own WebSocket connection. 

### Setup

To operate, each `VantiqSourceConnection` must be configured with a set of handlers.  These handlers configured as follows.  (Note that a similar call can be made using the `VantiqConnectorSet` if all connections should use the same handler.

Call `VantiqSourceConnection.configure_handlers(close_handler, connect_handler, publish_handler, query_handler)`.  These handlers are defined as follows.

* handle_close : Callable[[dict], Awaitable[None]]
* handle_connect : Callable[[dict, dict], Awaitable[None]
* handle_publish : Callable[[dict, dict], Awaitable[None]
* handle_query : Callable[[dict, dict], Awaitable[None]

> These parameters specify the callbacks to handle close, connect, publish, and query calls, respectively. For each, the first parameter is a dict providing the context for the operation.  The context will contain an entry keyed with 'source_name' whose value is the name of the source for which this is a callback.  For query calls, there will be an additional entry keyed with 'response_address'. The value here is used to route the query response back to the querier in the Vantiq server.

Reviewer: You should probably use the `VantiqConnector` constants here, instead of their values.
```suggestion
These parameters specify the callbacks to handle close, connect, publish, and query calls, respectively. For each, the first parameter is a dict providing the context for the operation.  The context will contain an entry keyed with `VantiqConnector.SOURCE_NAME` whose value is the name of the source for which this is a callback.  For query calls, there will be an additional entry keyed with `VantiqConnector.RESPONSE_ADDRESS`. The value here is used to route the query response back to the querier in the Vantiq server.
```

##### Reconnection
Under certain circumstances (_e.g._, a change to the source's configuration document or some network connection issue), the Vantiq server may disconnect
you from the source with the intent that you will immediately reconnect.

This operation is unavoidable, but is generally transparent to the connector.  When this happens, the [`handle_close` callback](#closeHandler) will be called, and then the
[`handle_connect` callback](#closeHandler).  Since this may be an indication that the source configuration has changed, each _connect_ call should be treated as a new connection/configuration.

This process is automatic.  No special work by the connector is required.

### Processing Messages

Messages involving a source come in three (3) flavors:

* Messages (events) sent by a source
* Query messages sent to a source. Query messages, in turn, have responses.
* Messages published to a source

The following sections outline the work involved here.

#### Notifications

Notifications are messages that the source will pass on to Vantiq rules that include 
`WHEN EVENT OCCURS ON "/sources/<source name>"`. To send one, simply call

> `Vantiq.send_notification(<message to be sent>)`, which will translate the message into JSON and add everything Vantiq needs to recognize the message.  The _message to be sent_ shoud be a `dict` object.

Reviewer: Should this be `VantiqSourceConnection.send_notification(<message to be sent>)`?

#### Queries

Queries are sent to the source via VAIL code of the form

```
row = SELECT * FROM SOURCE someSource with prop1 = value1, prop2 = value2, ...
```

A query like this will result in a query message to the source (and, thus, to the connector via the
`handle_query` callback). The parameter value passed to the callback will be a dict object
`{ prop1: value1, prop2: value2, ... }`.  That is, the parameter values are those passed from the _with_ clause in the VAIL SELECT statement.

Once the query message is received (_i.e._, delivered to the callback), the connector to which the
query was delivered must respond. It responds by making calls to return a query response or a query error.

##### Query Context

Query responses are responses to a `SELECT` request from Vantiq that targets a source, and can either be a `dict` or a series of `dict` objects in the form of a `list` or `array`. They only mean anything in relation to an initial Query message received from Vantiq, and thus should
only be sent as part response to a query via the callback. 

> To enable this behavior, the callback is passed a _context_ parameter that provides information about the context of the query. This parameter will take the form of a `dict` object with two (2)
> items: a `source_name` item that contains a string specifying the source toward which this query is
> sent, and a `response_address` item that contains the routing information to allow the query resposnes (or error) to find their way back the the Vantiq code that issued the query.

Reviewer: ```suggestion
To enable this behavior, the callback is passed a _context_ parameter that provides information about the context of the query. This parameter will take the form of a `dict` object with two (2)
items: the `VantiqConnector.SOURCE_NAME` item that contains a string specifying the source toward which this query is
sent, and the `VantiqConnector.RESPONSE_ADDRESS` item that contains the routing information to allow the query responses (or error) to find their way back the the Vantiq code that issued the query.
```

```

The `ctx` or context parameter is passed in to the `handle_query` callback.  Simply pass that back unchanged.

The code is an HTTP status code. You can use the following:

* `VantiqConnector.QUERY_EMPTY` to indicate that this query has an empty response.  In this case, the message parameter must be None.
* `VantiqConnector.QUERY_PARTIAL` to indicate that this is one of a number of responses to this query. This is used if you are dynamically producing results to return.
* `VantiqConnector.QUERY_COMPLETE` to indicate that this is the last (or only) response to this query.

For the latter two (2) codes, the `message` parameter must be either a `dict` or a `list` of `dict` objects.  These will be sent to the Vantiq server in a JSON format.

##### <a name="queryError" id="queryError"></a>Query Errors

Query errors are sent when a Query cannot be completed successfully. To send a Query error, call
`VantiqSourceConnection.send_query_error(ctx, message)`. 

The `ctx` or query context was passed in to the `handle_query` callback.
The `message` (`dict`) parameter must contain the following items.

* `VantiqConnector.ERROR_CODE` -- the short name for the error

> * `VantiqConnector.ERROR_TEMPLATE` -- template for there error message.  Templates contain zero0-based indexed references for parameters list (_e.g._,`{0}` to refer to the first parameter, `{1}` the second, etc.).

Reviewer: ```suggestion
* `VantiqConnector.ERROR_TEMPLATE` -- template for there error message.  Templates contain zero(0)-based indexed references for parameters list (_e.g._,`{0}` to refer to the first parameter, `{1}` the second, etc.).
```

* `VantiqConnector.ERROR_PARAMETERS` -- a list of parameters for the template.

##### <a name="publish" id="publish"></a>Published Messsages

When the Vantiq server publishes an event to a source that is implemented using a connector, it is delivered via the `handle_publish` callback. As with a query message, the handler will be passed two (2) parameter:  A context and the message.

> The context in this case will contain the `source_name` item.  It need not contain the `response_address` item.  You can not respond to messages delivered via the `handle_publish` callback.

Reviewer: ```suggestion
The context in this case will contain the `VantiqConnector.SOURCE_NAME` item.  It will not contain the `VantiqConnector.RESPONSE_ADDRESS` item.  You can not respond to messages delivered via the `handle_publish` callback.
```

If the Vantiq server executes VAIL code like

```
    PUBLISH { prop1: value1, prop2: value2, ... } TO SOURCE connectorSource
```

the connector will receive a `handle_callback` call containing the context parameter

```python
{ 
    'source_name': 'connectorSource'
}
```

and a message parameter containing

```python
{ 
    'prop1': value1,
    'prop2': value2
}
```

The connector can then act on the message as it deems appropriate.

##### <a name="closeHandler" id="closeHandler"></a>Close
The closure handler does not deal with a specific message or type of message, but instead is called when either your
code calls `VantiqSourceConnection.close()` or the WebSocket connection is forced to close, most likely due to a problem with the connection.

> The `handle_close` is passed a context containing the source name (see [publish](#publish)). At the time the handler is called, the connection has already been closed.  No interaction with the Vantiq server is possible;  this handler call allows the connector to perform an resource deallocation or other cleanup necessary.

Reviewer: ```suggestion
The `handle_close` is passed a context containing the source name (see [publish](#publish)). At the time the handler is called, the connection has already been closed.  No interaction with the Vantiq server is possible;  this handler call allows the connector to perform any resource deallocation or other cleanup necessary.
```
